scylla-cql/src/frame/value.rs

```diff
@@ -906,17 +1033,78 @@ impl_batch_values_for_tuple!(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T
                              0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15; 16);
 
 // Every &impl BatchValues should also implement BatchValues
-impl<T: BatchValues> BatchValues for &T {
-    fn len(&self) -> usize {
-        <T as BatchValues>::len(*self)
+impl<'a, 'r, T: BatchValues + ?Sized> BatchValuesGatWorkaround<'r> for &'a T {
+    type BatchValuesIter = <T as BatchValuesGatWorkaround<'a>>::BatchValuesIter;
+    fn batch_values_iter(&'r self) -> Self::BatchValuesIter {
+        <T as BatchValuesGatWorkaround<'a>>::batch_values_iter(*self)
     }
+    fn len(&'r self) -> usize {
+        <T as BatchValuesGatWorkaround<'a>>::len(*self)
+    }
+}
 
-    fn write_nth_to_request(
-        &self,
-        n: usize,
+/// Allows reusing already-serialized first value
+///
+/// We'll need to build a `SerializedValues` for the first ~`ValueList` of a batch to figure out the shard (#448).
+/// Once that is done, we can use that instead of re-serializing.
+///
+/// This struct implements both `BatchValues` and `BatchValuesIterator` for that purpose
+pub struct BatchValuesWithMaybeFirstAlreadySerializedValue<'f, T> {
+    first: Option<&'f SerializedValues>,
+    rest: T,
+}
+impl<'f, T: BatchValues> BatchValuesWithMaybeFirstAlreadySerializedValue<'f, T> {
+    pub fn new(batch_values: T, already_serialized_first: Option<&'f SerializedValues>) -> Self {
+        Self {
+            first: already_serialized_first,
+            rest: batch_values,
+        }
+    }
+}
+impl<'r, 'f, BV: BatchValues> BatchValuesGatWorkaround<'r>
+    for BatchValuesWithMaybeFirstAlreadySerializedValue<'f, BV>
+{
+    type BatchValuesIter = BatchValuesWithMaybeFirstAlreadySerializedValue<
+        'f,
+        <BV as BatchValuesGatWorkaround<'r>>::BatchValuesIter,
+    >;
+    fn batch_values_iter(&'r self) -> Self::BatchValuesIter {
+        BatchValuesWithMaybeFirstAlreadySerializedValue {
+            first: self.first,
+            rest: self.rest.batch_values_iter(),
+        }
+    }
+    fn len(&'r self) -> usize {
+        self.rest.len()
+    }
+}
+impl<'f, IT: BatchValuesIterator> BatchValuesIterator
+    for BatchValuesWithMaybeFirstAlreadySerializedValue<'f, IT>
+{
+    fn next_serialized(&mut self) -> Option<SerializedResult<'_>> {
+        match self.first.take() {
+            Some(first) => {
+                self.rest.skip_next();
+                Some(Ok(Cow::Borrowed(first)))
+            }
+            None => self.rest.next_serialized(),
+        }
+    }
+    fn write_next_to_request(
+        &mut self,
         buf: &mut impl BufMut,
-    ) -> Result<(), SerializeValuesError> {
-        <T as BatchValues>::write_nth_to_request(*self, n, buf)?;
-        Ok(())
+    ) -> Option<Result<(), SerializeValuesError>> {
+        match self.first.take() {
+            Some(first) => {
+                self.rest.skip_next();
+                #[allow(clippy::unit_arg)]
+                Some(Ok(first.write_to_request(buf)))
```

Comment:
I agree with clippy here.

```rust
first.write_to_request(buf);
Some(Ok(()))
```
Is more readable.

We try to listen to clippy whenever possible, it reduces strange `#[allow` lines and removes the need for discussing style choices.

Reply:
My idea is that if the return type of `write_to_request` becomes something else, we may want to consider whether to return it there, and the compilation error that would appear would allow that.
With the alternate writing suggested by clippy, that wouldn't happen.

Anyway, I'm not too attached to it so done.